File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> outer = new ArrayList<>();
        List<Integer> comb = new ArrayList<>();
        sum(candidates,0,target,outer,comb);
        return outer;
        
    }
    private static void sum(int[] arr,int index,int target,List<List<Integer>> outer,List<Integer> comb){
        if(target==0){
            if(!outer.contains(comb)){
                outer.add(new ArrayList<>(comb));
            }
             
            return;
        }
        if(index==arr.length||target<0){
            return;
        }
        
        comb.add(arr[index]);
        sum(arr,index+1,target-arr[index],outer,comb);
        sum(arr,index,target-arr[index],outer,comb);
        comb.remove(comb.size()-1);
        sum(arr,index+1,target,outer,comb);
    }
}

File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    vector<vector<int>> result;
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> temp;
        soln(candidates,target,temp,0,0);
        return result;
    }

    void soln(vector<int> nums,int target,vector<int> temp,int sum,int index){
        if(sum>target){
            return;
        }
        if(sum==target){
            result.push_back(temp);
            return;
        }
        for(int i=index;i<nums.size();i++){
            sum+=nums[i];
            temp.push_back(nums[i]);
            soln(nums,target,temp,sum,i);
            temp.pop_back();
            sum-=nums[i];
        }
    }
};